### PR TITLE
Fix MoE weight loading

### PR DESCRIPTION
**Pull request by mikasenghaas**
This PR fixes a weight loading issue that occurs when we try to load a custom (MoE) model (`--model.impl custom`) from meta device (`--model.load-using-meta`). Our custom model uses the TT format for MoE layers. When trying to load a HF MoE model (e.g. `zai-org/GLM-4.5-Air`) the FQNs do not match leading to wrongly loading the weights. Because we had `DefaultLoadPlanner(allow_partial_load=True)` this would error siltently and would manifest in high loss on the SFT trainer and high trainer-inference mismatch on the RL trainer.

Changes made:
- Use the default planner (`allow_partial_load=False`) to get explicit logs
- Automatically detect and convert if there is a mismatch between the snapshot weights and the model weights and do a best-effort attempt at conversion. This only happens on the master rank (all non-master ranks wait), and is "cached", i.e. we write the converted weights to disk for subsequent operations.

Note: The bug will still occur if one tries to load with `--model.impl custom` on CPU device.

### CODE REVIEW

File: src/prime_rl/trainer/model.py [hunk 1]
```diff
@@ -152,35 +163,68 @@ def setup_fsdp(model: nn.Module, config: ModelConfig, parallel_dims: ParallelDim
...
+
+    logger.info(f"Loading weights using HF DCP from {snapshot_path}")
+    load_dcp_start_time = time.time()
+    dcp_load(
         model.state_dict(),
-        storage_reader=HuggingFaceStorageReader(path=path_snapshot),
+        storage_reader=HuggingFaceStorageReader(path=snapshot_path.as_posix()),
         # Note: This allow is needed by weight tying but could cause silent issues
-        planner=DefaultLoadPlanner(allow_partial_load=True),
+        # planner=DefaultLoadPlanner(allow_partial_load=True),
```
Comment (samsja) [inline src/prime_rl/trainer/model.py:224]: why commented out, btw should we put allow_partial_load to false here

File: src/prime_rl/trainer/weights.py [hunk 2]
```diff
@@ -99,6 +161,55 @@ def _convert_tt_moe_to_hf_(state_dict: dict[str, Tensor]):
         del state_dict[f"model.layers.{i}.mlp.experts.w3"]
 
 
+def load_state_dict(path: Path) -> dict[str, Tensor]:
```
Comment (samsja) [inline src/prime_rl/trainer/weights.py]:
```suggestion
def load_state_dict_safetensor(path: Path) -> dict[str, Tensor]:
```

File: src/prime_rl/trainer/weights.py [hunk 3]
```diff
@@ -99,6 +161,55 @@ def _convert_tt_moe_to_hf_(state_dict: dict[str, Tensor]):
...
+    safetensors_paths.sort(key=lambda x: int(x.stem.split("-")[1].split("of")[0]))
+    state_dict = {}
+    for safetensor_path in tqdm(safetensors_paths, desc="Loading weight shards", total=len(safetensors_paths)):
+        with safe_open(safetensor_path, framework="pt", device="cpu") as f:
+            for key in f.keys():
+                state_dict[key] = f.get_tensor(key)
+    return state_dict
+
+
+def save_state_dict(state_dict: dict[str, Tensor], path: Path):
```
Comment (samsja) [inline src/prime_rl/trainer/weights.py]:
```suggestion
def save_state_dict_safetensor(state_dict: dict[str, Tensor], path: Path):
```

File: src/prime_rl/trainer/model.py (hunk 1 above)
Comment (mikasenghaas) [reply src/prime_rl/trainer/model.py:224]: commented out bc i wanted to get your opinion here. given jack's comments it seems like this is needed for some models? but also, i really don't like allowing partial load because it fails siltently. disabling this showed me exactly which keys in the state dict were not matching which is a much nicer debugging experience.
